src/main/java/org/variantsync/diffdetective/diff/result/DiffError.java

```diff
@@ -1,61 +1,81 @@
      */
-    public final static DiffError COMMIT_HAS_NO_PARENTS = new DiffError("commit has no parents");
+    COMMIT_HAS_NO_PARENTS("commit has no parents"),
 
     /**
      * Some internal error occurred when operating JGit.
      * The error is not further specified or unknown.
      */
-    public final static DiffError JGIT_ERROR = new DiffError("error when operating jgit");
+    JGIT_ERROR("error when operating jgit"),
 
     /**
      * An error which occurred when obtaining the full diff from a local diff.
      */
-    public final static DiffError COULD_NOT_OBTAIN_FULLDIFF = new DiffError("could not obtain full diff");
+    COULD_NOT_OBTAIN_FULLDIFF("could not obtain full diff"),
 
     /**
      * A preprocessor block was opened but not closed.
      * Typically, this occurs when an #endif is missing.
      */
-    public final static DiffError NOT_ALL_ANNOTATIONS_CLOSED = new DiffError("not all annotations closed");
+    NOT_ALL_ANNOTATIONS_CLOSED("not all annotations closed"),
 
     /**
      * A file or patch contained an expression that closes an annotation block (typically an #endif)
      * but there is no block to close.
      */
-    public final static DiffError ENDIF_WITHOUT_IF = new DiffError("#endif without #if");
+    ENDIF_WITHOUT_IF("#endif without #if"),
 
     /**
      * A multiline macro was defined within a multiline macro.
      */
-    public final static DiffError MLMACRO_WITHIN_MLMACRO = new DiffError("definition of multiline macro within multiline macro");
+    MLMACRO_WITHIN_MLMACRO("definition of multiline macro within multiline macro"),
 
     /**
      * An #else or #elif expression has no corresponding #if (or #ifdef, ...) expression.
      */
-    public final static DiffError ELSE_OR_ELIF_WITHOUT_IF = new DiffError("#else or #elif without #if");
+    ELSE_OR_ELIF_WITHOUT_IF("#else or #elif without #if"),
 
     /**
      * An #else expression is followed by another #else expression which has no semantics.
      */
-    public final static DiffError ELSE_AFTER_ELSE = new DiffError("#else after #else");
+    ELSE_AFTER_ELSE("#else after #else"),
 
     /**
      * A condition annotation is missing an expression.
      * This typically occurs when an #if macro has no arguments.
      */
-    public final static DiffError IF_WITHOUT_CONDITION = new DiffError("conditional macro without expression");
+    IF_WITHOUT_CONDITION("conditional macro without expression");
+
+    private final String message;
+
+    private DiffError(String message) {
+        this.message = message;
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+    public static Optional<DiffError> fromMessage(String message) {
+        return Arrays
+            .stream(values())
+            .filter(m -> m.getMessage().equals(message))
+            .findFirst();
+    }
 
     @Override
     public String toString() {
-        return id;
+        return message;
     }
 }
```

Comment:
In my opinion all errors produces by the code in DiffDetective should be added to such an enum. If there is someone who wants to handle some errors, but not all, he can differentiate between these errors without relying on a string which is probably not really public API. The compiler can also provide errors if the error name is changed or doesn't exist any more.
If the users of DiffDetective as a library should be able to add new error codes, it might be a better idea to introduce an interface for such error codes and let the enum implement this interface. Although I currently don't know of a use case where users of DiffDetective should be able to pass custom errors into DiffDetective.

Reply:
I agree. The current solution is probably the best and definitely a good one for our use cases so far.